db/bigtable_eth1.go

```diff
@@ -3594,6 +3734,184 @@ func (bigtable *Bigtable) GetAddressNames(addresses map[string]string) error {
 	return err
 }
 
+type isContractInfo struct {
+	update *types.IsContractUpdate
+	ts     gcp_bigtable.Timestamp
+}
+
+type isContractAtRequest struct {
+	address  string
+	block    int64
+	txIdx    int64
+	traceIdx int64
+}
+
+func (bigtable *Bigtable) getAddressIsContractHistories(histories map[string][]isContractInfo) error {
+	if len(histories) == 0 {
+		return nil
+	}
+
+	tmr := time.AfterFunc(REPORT_TIMEOUT, func() {
+		logger.WithFields(logrus.Fields{}).Warnf("%s call took longer than %v", utils.GetCurrentFuncName(), REPORT_TIMEOUT)
+	})
+	defer tmr.Stop()
+
+	keys := make([]string, 0, len(histories))
+	for address := range histories {
+		keys = append(keys, fmt.Sprintf("%s:%s", bigtable.chainId, address))
+	}
+
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(time.Second*30))
+	defer cancel()
+
+	filter := gcp_bigtable.ChainFilters(gcp_bigtable.FamilyFilter(ACCOUNT_METADATA_FAMILY), gcp_bigtable.ColumnFilter(ACCOUNT_IS_CONTRACT))
+
+	keyPrefix := fmt.Sprintf("%s:", bigtable.chainId)
+	err := bigtable.tableMetadata.ReadRows(ctx, gcp_bigtable.RowList(keys), func(row gcp_bigtable.Row) bool {
+		address := strings.TrimPrefix(row.Key(), keyPrefix)
+		for _, v := range row[ACCOUNT_METADATA_FAMILY] {
+			b := &types.IsContractUpdate{}
+			err := proto.Unmarshal(v.Value, b)
+			if err != nil {
+				utils.LogError(err, "error parsing IsContractUpdate data", 0)
+			}
+			histories[address] = append(histories[address], isContractInfo{update: b, ts: v.Timestamp})
+		}
+
+		return true
+	}, gcp_bigtable.RowFilter(filter))
+
+	if err != nil {
+		return fmt.Errorf("error reading isContract histories from bigtable: %w", err)
+	}
+
+	return nil
+}
+
+// returns account state after the given execution state
+// -1 is latest (e.g. "txIdx" = -1 returns the contract state after execution of "block", "block" = -1 returns the state at chain head)
+func (bigtable *Bigtable) GetAddressIsContractAt(requests []isContractAtRequest) ([]types.ContractInteractionType, error) {
+	results := make([]types.ContractInteractionType, len(requests))
+	if len(requests) == 0 {
+		return results, nil
+	}
+
+	// get histories
+	histories := make(map[string][]isContractInfo, 0)
+	for _, request := range requests {
+		histories[request.address] = nil
+	}
+	err := bigtable.getAddressIsContractHistories(histories)
+	if err != nil {
+		return nil, err
+	}
+
+	// evaluate requests; CONTRACT_NONE is default
+	for i, request := range requests {
+		if _, ok := histories[request.address]; !ok {
+			continue
+		}
+		if histories[request.address] == nil {
+			continue
+		}
+		history := histories[request.address]
```

Comment:
**Nitpick:** Why not write
```suggestion
		if history, ok := histories[request.address]; !ok {
			continue
		}
		if history == nil {
			continue
		}
```
?

Reply:
Well, to nitpick on the nitpick, because variables are only defined in the scope in which they are declared :stuck_out_tongue: so `history` wouldn't be known after the first `if`. But by moving the if I can still simplify this a lot, thanks